state/virtualhostkeys.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/juju/core/virtualhostkeys"
+	"github.com/juju/mgo/v3"
+	"github.com/juju/mgo/v3/txn"
+)
+
+// unitHostKeyID provides the virtual host key
+// lookup value for a unit based on the unit name.
+func unitHostKeyID(unitName string) string {
+	return "unit" + "-" + unitName + "-" + "hostkey"
+}
+
+// machineHostKeyID provides the virtual host key
+// lookup value for a machine based on the machine ID.
+func machineHostKeyID(machineID string) string {
+	return "machine" + "-" + machineID + "-" + "hostkey"
+}
+
+// VirtualHostKey represents the state of a virtual host key.
+type VirtualHostKey struct {
+	st  *State
+	doc virtualHostKeyDoc
+}
+
+type virtualHostKeyDoc struct {
+	DocId   string `bson:"_id"`
+	HostKey []byte `bson:"hostkey"`
+}
+
+// HostKey returns the virtual host key.
+func (s *VirtualHostKey) HostKey() []byte {
+	return s.doc.HostKey
+}
+
+func newVirtualHostKeyDoc(st *State, hostKeyID string, hostkey []byte) (virtualHostKeyDoc, error) {
+	return virtualHostKeyDoc{
+		DocId:   st.docID(hostKeyID),
+		HostKey: hostkey,
+	}, nil
+}
+
+func newMachineVirtualHostKeysOps(st *State, machineID string, hostKey []byte) ([]txn.Op, error) {
+	hostKeyID := virtualhostkeys.MachineHostKeyID(machineID)
+	doc, err := newVirtualHostKeyDoc(st, hostKeyID.ID, hostKey)
+	if err != nil {
+		return nil, err
+	}
+	return []txn.Op{{
+		C:      virtualHostKeysC,
+		Id:     doc.DocId,
+		Assert: txn.DocMissing,
+		Insert: doc,
+	}}, nil
+}
+
+func newUnitVirtualHostKeysOps(st *State, unitName string, hostKey []byte) ([]txn.Op, error) {
+	hostKeyID := virtualhostkeys.UnitHostKeyID(unitName)
+	doc, err := newVirtualHostKeyDoc(st, hostKeyID.ID, hostKey)
+	if err != nil {
+		return nil, err
+	}
+	return []txn.Op{{
+		C:      virtualHostKeysC,
+		Id:     doc.DocId,
+		Assert: txn.DocMissing,
+		Insert: doc,
+	}}, nil
+}
+
+func removeMachineVirtualHostKeyOps(state *State, machineID string) []txn.Op {
+	machineLookup := virtualhostkeys.MachineHostKeyID(machineID)
+	docID := state.docID(machineLookup.ID)
+	return []txn.Op{{
+		C:      virtualHostKeysC,
+		Id:     docID,
+		Remove: true,
+	}}
+}
+
+func removeUnitVirtualHostKeysOps(state *State, unitName string) []txn.Op {
+	unitLookup := unitHostKeyID(unitName)
+	docID := state.docID(unitLookup)
+	return []txn.Op{{
+		C:      virtualHostKeysC,
+		Id:     docID,
+		Remove: true,
+	}}
+}
+
+// MachineVirtualHostKey returns the virtual host key for a machine.
+func (st *State) MachineVirtualHostKey(machineID string) (*VirtualHostKey, error) {
+	return st.virtualHostKey(machineHostKeyID(machineID))
+}
+
+// UnitVirtualHostKey returns the virtual host key for a unit.
```

Comment:
For a CAAS unit?

Reply:
Leaving this as is because the logic that only stores virtual host keys for CAAS units is not defined here. Basically, it would be confusing to read "for a CAAS unit" but then nowhere in this method is it doing anything about enforcing it is for caas units.